helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -106,61 +115,131 @@ public RoutingTableProvider(HelixManager helixManager, PropertyType sourceDataTy
     _helixManager = helixManager;
-    _sourceDataType = sourceDataType;
+    _routerUpdater = new HashMap<>();
+    _sourceDataTypes = sourceDataTypes;
     _routingTableChangeListenerMap = new ConcurrentHashMap<>();
     String clusterName = _helixManager != null ? _helixManager.getClusterName() : null;
+    _monitor = new HashMap<>();
 
-    _monitor = new RoutingTableProviderMonitor(_sourceDataType, clusterName);
-    try {
-      _monitor.register();
-    } catch (JMException e) {
-      logger.error("Failed to register RoutingTableProvider monitor MBean.", e);
+    for (PropertyType propertyType : _sourceDataTypes.keySet()) {
+      _monitor.put(propertyType, new RoutingTableProviderMonitor(propertyType, clusterName));
+      try {
+        _monitor.get(propertyType).register();
+      } catch (JMException e) {
+        logger.error("Failed to register RoutingTableProvider monitor MBean.", e);
+      }
     }
     _reportExecutor = Executors.newSingleThreadExecutor();
 
-    _routerUpdater = new RouterUpdater(clusterName, _sourceDataType);
-    _routerUpdater.start();
-
-    if (_helixManager != null) {
-      switch (_sourceDataType) {
-      case EXTERNALVIEW:
-        try {
-          _helixManager.addExternalViewChangeListener(this);
-        } catch (Exception e) {
-          shutdown();
-          logger.error("Failed to attach ExternalView Listener to HelixManager!");
-          throw new HelixException("Failed to attach ExternalView Listener to HelixManager!", e);
-        }
-        break;
-
-      case TARGETEXTERNALVIEW:
-        // Check whether target external has been enabled or not
-        if (!_helixManager.getHelixDataAccessor().getBaseDataAccessor().exists(
-            _helixManager.getHelixDataAccessor().keyBuilder().targetExternalViews().getPath(), 0)) {
-          shutdown();
-          throw new HelixException("Target External View is not enabled!");
+    for (PropertyType propertyType : _sourceDataTypes.keySet()) {
+      if (_sourceDataTypes.get(propertyType).size() == 0) {
+        // Empty aggregationType
+        String aggregationType = "";
+        String key = propertyType.name() + "_" + aggregationType;
+        if (_routerUpdater.get(key) == null) {
+          _routerUpdater.put(key, new RouterUpdater(clusterName, propertyType, aggregationType, key));
+          _routerUpdater.get(key).start();
+          _routingTableRef.put(key, new AtomicReference<>(new RoutingTable(propertyType)));
         }
-
-        try {
-          _helixManager.addTargetExternalViewChangeListener(this);
-        } catch (Exception e) {
-          shutdown();
-          logger.error("Failed to attach TargetExternalView Listener to HelixManager!");
-          throw new HelixException("Failed to attach TargetExternalView Listener to HelixManager!",
-              e);
+      } else {
+        for (String aggregationType : _sourceDataTypes.get(propertyType)) {
+          String key = propertyType.name() + "_" + aggregationType;
```

Comment:
This can be combined to a method together with the previous condition's code. And, actually, I suggest to put this logic into routerUpdater. And make the routerUpdater support multiple types.

Reply:
I move as much as I could to the RouterUpdater. the key has been used to _routingTableRefMap now here. Now routerUpdater supports multiple types.